abi/src/main/java/org/web3j/abi/TypeDecoder.java

```diff
@@ -464,7 +475,92 @@ public static <T extends Type> T decodeDynamicStruct(
                     }
                 };
 
-        return decodeDynamicStructElements(input, offset, typeReference, function);
+        if (typeReference.getClassType().isAssignableFrom(DynamicStruct.class)
+                && typeReference.getInnerTypes() != null) {
+            return decodeDynamicStructElementsFromInnerTypes(
+                    input, offset, typeReference, function);
+        } else {
+            return decodeDynamicStructElements(input, offset, typeReference, function);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T extends Type> T decodeDynamicStructElementsFromInnerTypes(
+            final String input,
+            final int offset,
+            final TypeReference<T> typeReference,
+            final BiFunction<List<T>, String, T> consumer)
+            throws ClassNotFoundException {
+        final Class<T> classType = typeReference.getClassType();
+        final List<TypeReference<?>> innerTypes = typeReference.getInnerTypes();
+        final int length = innerTypes.size();
+        final Map<Integer, T> parameters = new HashMap<>();
+        int staticOffset = 0;
+        final List<Integer> parameterOffsets = new ArrayList<>();
+
+        int dynamicParametersToProcess = 0;
+        for (int i = 0; i < length; ++i) {
```

Comment:
extract this for in a separated method to be easier to follow as this is computing the offsets

Reply:
Change has been made in  https://github.com/hyperledger/web3j/pull/2076/commits/a08ac8470d59272a4f6bf9e43daa9fe398555720.